### DISCUSSION TITLE
Understanding of rate limiting and retry behavior in Crossplane controllers

### DISCUSSION THREAD

**Discussion by lindblombr** [crossplane/crossplane#6079] [Q&A] [answered]
Hi folks! We have a provider implemented on crossplane-runtime.  This obviously involves some interaction with controller-runtime and this may be a better question for that community, but since this is a crossplane provider and it is specifically geared towards management of resources external to the cluster, I figured folks here would have a better grasp :)

## Background

Our provider has worked reasonably well for us with a fairly bog-standard controller definition like so

```go
        r := reconciler...
	return ctrl.NewControllerManagedBy(mgr).
		Named(name).
		WithOptions(o.ForControllerRuntime()).
		For(&v1.MyManagedResource{}).
		Complete(ratelimiter.NewReconciler(name, r, o.GlobalRateLimiter))
```

In practice, we've found that we need to back-off much more aggressively than the standard rate limiter supplied within `o.ForControllerRuntime()` (which seems to be a fairly common configuration for providers) that specifies the rate limiter thusly

```go
// ForControllerRuntime extracts options for controller-runtime.
func (o Options) ForControllerRuntime() controller.Options {
	recoverPanic := true

	return controller.Options{
		MaxConcurrentReconciles: o.MaxConcurrentReconciles,
		RateLimiter:             ratelimiter.NewController(),
		RecoverPanic:            &recoverPanic,
	}
}
```

and

```go
// NewController returns a rate limiter that takes the maximum delay between the
// passed rate limiter and a per-item exponential backoff limiter. The
// exponential backoff limiter has a base delay of 1s and a maximum of 60s.
func NewController() ratelimiter.RateLimiter {
	return workqueue.NewItemExponentialFailureRateLimiter(1*time.Second, 60*time.Second)
}
```

The initial thought here was to modify our controller builder definition like so

_Naive Change_
```go
        r := reconciler...
        rateLimiter := workqueue.NewItemExponentialFailureRateLimiter(2*time.Minute, 15*time.Minute)
        controllerOptions := o.ForControllerRuntime()
        controllerOptions.RateLimiter = rateLimiter
	return ctrl.NewControllerManagedBy(mgr).
		Named(name).
		WithOptions(controllerOptions).
		For(&v1.MyManagedResource{}).
		Complete(ratelimiter.NewReconciler(name, r, o.GlobalRateLimiter))
```

You can see here that we've backed off MUCH more aggressively than the standard rate limiter.

When we make this change, however, we do not observe the desired backoff behavior.  Instead, requests are retried as quickly as they are re-queued and within the bounds of `--max-reconcile-rate`.

I then wondered: why are we passing a rate limiter via WithOptions, then wrapping the reconciler in the GlobalRateLimiter?  Why would we not do something like this, instead, for clarity:

_Proposed, Working Approach_
```go
        r := reconciler...
        rateLimiter := workqueue.NewMaxOfRateLimiter(
		workqueue.NewItemExponentialFailureRateLimiter(2*time.Minute, 15*time.Minute),
		// Honor the global rate limiter
		o.GlobalRateLimiter,
	)
        controllerOptions := o.ForControllerRuntime()
        controllerOptions.RateLimiter = rateLimiter
	return ctrl.NewControllerManagedBy(mgr).
		Named(name).
		WithOptions(controllerOptions).
		For(&v1.MyManagedResource{}).
		Complete(r)
```

With this, I achieve the desired behavior:
* Resources reconciled according to poll interval
* Failed reconciliations being retried with a minimum delay of 2 minutes between retries

Since the previous pattern appears so pervasive in practice...

See
* https://github.com/crossplane-contrib/provider-kubernetes/blob/7ba46755740bc29f61f05c9251f248317a2bd6ae/internal/controller/object/object.go#L239
* https://github.com/crossplane-contrib/provider-upjet-aws/blob/cc7b958eb41f4466d707cedf8862d8369e15e43c/internal/controller/rds/cluster/zz_controller.go#L94

I'm sitting here wondering... what am I missing in my understanding here?  Why did my _Naive Approach_ not yield the desired results? What is problematic with my _Proposed, Working Approach_?  What purpose does wrapping the reconciler in the global rate limiter serve that a MaxOfRateLimiter doesn't?  

Other details

* `github.com/crossplane/crossplane-runtime v1.17.0`
* `sigs.k8s.io/controller-runtime v0.18.2`

Thanks in advance

**Comment by lindblombr** [comment]
Also, [this](https://github.com/crossplane/crossplane-runtime/blob/19d95a69cc03690c4b867ff91d89681fcf872a93/pkg/ratelimiter/default.go#L39) code comment

```go
// NewController returns a rate limiter that takes the maximum delay between the
// passed rate limiter and a per-item exponential backoff limiter. The
// exponential backoff limiter has a base delay of 1s and a maximum of 60s.
func NewController() ControllerRateLimiter {
	return workqueue.NewTypedItemExponentialFailureRateLimiter[reconcile.Request](1*time.Second, 60*time.Second)
}
```

"returns a rate limiter that takes the maximum delay between the *passed* rate limiter..."

Where is this "*passed* rate limiter actually _passed_?

**Comment by jbw976** [comment]
@lindblombr I didn't read through and understand the details of this post yet, but just quickly wanted to confirm that you already saw the retroactive 1-pager that @negz added for rate limiting within the last year?  That could have some relevant insight for you if you haven't already seen it (but you probably already have 😉)

https://github.com/crossplane/crossplane/blob/main/design/one-pager-rate-limiting.md

**Reply by lindblombr** [reply]
Thanks @jbw976 .  Basically, the document aligns with the initial implementation and what I attempted to achieve with my _Naive Change_ .  I'm wondering if there's some dependency version mismatch since, based on the doc, I would expect the _Naive Change_ to produce the desired behavior

**Comment by ravilr** [comment]
>  What is problematic with my Proposed, Working Approach? What purpose does wrapping the reconciler in the global rate limiter serve that a MaxOfRateLimiter doesn't?

> https://github.com/crossplane/crossplane/blob/main/design/one-pager-rate-limiting.md

Above notes by @negz  already lays out the difference.   There are 4 reasons/triggers a reconcile request would be enqueued into the workqueue:
1.) Actual watch event for the Object, due to desired spec change/generation change : Not subject to controller-runtime's RateLimiter.
2.) Reconcile returning RequeueAfter: T.time, every poll interval, leading to enqueue :  Not subject to controller-runtime's RateLimiter.
3.) Reconcile returning RequeueAfter: true for immediate retries, leading to enqueue : Subject to controller-runtime's RateLimiter.
4.) Reconcile returning error, leading to enqueue: : Subject to controller-runtime's RateLimiter.

the controller-runtime's RateLimiter set in controller.Options is only applicable for last/latter 2 triggers above: https://github.com/kubernetes-sigs/controller-runtime/blob/v0.18.2/pkg/internal/controller/controller.go#L312-#L344

@negz the notes in the docs seems to mention, even the reconcile requests enqueues triggered by actual watch event are subjected to controller-runtime's RateLimiter.  They aren't right ?

Which means, for actual watch events and periodic poll interval enqueued requests get dequeued and processed as soon as possible, which means say 100 MRs getting spec updates at once/same time would result in 100 reconcile requests enqueued immediately, without any ratelimiting, which with enough workers(concurrent workers) configured can bombard the external provider API at the same time/concurrently.

That's where the crossplane-runtime's Reconciler middleware wrapping the inner reconciler and enforcing the GlobalRateLimiter comes into play, so that all reconcile triggers are ratelimited, not just on Reconcile returning error/Requeue:true. 

[crossplane-runtime/pkg/ratelimiter default NewController()](https://github.com/crossplane/crossplane-runtime/blob/v1.17.0/pkg/ratelimiter/default.go#L39) ratelmiter is passed onto controller-runtime's RateLimiter Option, as a per-item exponential backoff limiter to rate limit reconciles that return Requeue: true or a non-nil error.

[crossplane-runtime/pkg/ratelimiter default NewGlobal()](https://github.com/crossplane/crossplane-runtime/blob/v1.17.0/pkg/ratelimiter/default.go#L32) ratelmiter is passed to crossplane-runtime's ratelimiter.NewReconciler middleware wrapper to rate limit all reconciles.

**Comment by ravilr** [comment, marked as answer]
> Why did my Naive Approach not yield the desired results? 

Are you able to verify the source of the trigger(out of above 4) for the reconcile requests for the same Object(s) being processed immediately again and appearing to be retries ?

provider controllers are recommended to configure a event filter in the controller Setup to filter out the `.Status` updates triggering reconcile request loop using the [DesiredStateChanged predicate](https://github.com/crossplane/crossplane-runtime/blob/v1.17.0/pkg/resource/predicates.go#L45).   xref https://github.com/crossplane/crossplane/issues/3455 on how it could impact, without that event filter predicate.

**Comment by ravilr** [comment]
@negz on a related note, the crossplane-core controllers(claim/offered, defintion/composite, composition) don't seem to have DesiredStateChanged predicate filter configured to filter out status updates, and hence suffers from endless reconcile requests loop, on error returned by reconcile() -> status subresource update -> watch event -> reconcile() -> repeat loop.., without any ratelimiting backoff.
below is a screenshot of kube-apiserver audit log events showing the problematic issue on a crossplane instance with few composite resources in error state, and that resulting in 100+ per second, status subresource update kube-api API calls from defined and offered reconcilers:

<img width="1452" alt="Screenshot 2024-11-08 at 12 34 21 AM" src="https://github.com/user-attachments/assets/6935e967-bdea-4462-9f0b-2e7227c4ad15">

**Comment by lindblombr** [comment]
@ravilr This may be exactly what the issue is.  There is some status information updated on Observe that is likely falling into the first category.  I'm experimenting with adding the event filter with the DesiredStateChange predicate and will report back.

**Comment by lindblombr** [comment]
@ravilr That was _exactly_ the issue in this case.  Thank you!  After including the DesiredStateChange predicate as an event filter, I am observing the desired backoff delays.

The following code snippet gives the desired behavior while still being able to update `status` information (rates are a bit different from original snippet)

```go
        r := reconciler...
        rateLimiter := workqueue.NewItemExponentialFailureRateLimiter(30*time.Second, 2*time.Minute)
        controllerOptions := o.ForControllerRuntime()
        controllerOptions.RateLimiter = rateLimiter
	return ctrl.NewControllerManagedBy(mgr).
		Named(name).
		WithOptions(controllerOptions).
                WithEventFilter(resource.DesiredStateChanged()).  // Ensure status changes don't bypass rate limits
		For(&v1.MyManagedResource{}).
		Complete(ratelimiter.NewReconciler(name, r, o.GlobalRateLimiter))
```

**Comment by bobh66** [comment]
This makes me wonder if we don't need the same sort of filter on core crossplane composite resources.  I am seeing the Crossplane pod spinning on the same composite resource continuously with no backoff which sounds an awful lot like this problem.  Maybe I'll try this fix there and see if it solves that problem.  Thanks Brian!

**Reply by lindblombr** [reply]
I give credit to @ravilr for calling out the core controller part :)

Is there a canonical "provider developer" guide that may benefit from expanding on backoff behavior documentation that I can maybe contribute to?  So someone else doesn't bump into this later?

**Comment by bobh66** [comment]
@negz should the `WithEventFilter(resource.DesiredStateChanged())` be applied to core Crossplane controllers too?

**Comment by lindblombr** [comment]
Another part to this that I thought about...

None of the providers allow for configuration of the backoff duration and maximum.  Nearly all of them use the defaults provided by `ratelimite.NewController`, which is better than the controller-runtime defaults, but not appropriate for all situations.  provider-terraform is a great example of something that probably has little business re-trying a reconciliation after 1 second, up to every 15 seconds.  I know this has been discussed in the past, but it seems like providing configurability here is valuable, especially since there are external cost implications involved.  I'd even go so far as to say that it may be preferable that this behavior be _explicitly_ declared for each provider and that the provider implementations should allow for user configuration of those values based on their needs.  This encourages provider authors to really think about the most appropriate values for this configuration.
